testing/G_controllers.md

```diff
@@ -1 +1,297 @@
+  end
+
+  def render("user.json", %{user: user}) do
+    %{name: user.name, email: user.email}
+  end
+
+end
+```
+
+And with that, our test passes when we run it.
+
+We'll also cover the `show/2` action here so we can see how to handle an error case.
+
+Our show tests currently look like this:
+
+```elixir
+  describe "show/2" do
+    test "Responds with a newly created user if the user is found"
+    test "Responds with a message indicating user not found"
+  end
+```
+
+Run this test only by running the following command: (if your show tests don't start on line 32, change the line number accordingly)
+
+```bash
+$ mix test test/controllers/user_controller_test.exs:32
+```
+
+Our first `show/2` test result is, as expected, not yet implemented.
+Let's build a test around what we think a successful `show/2` should look like.
+
+```elixir
+test "Reponds with a newly created user if the user is found", %{conn: conn} do
+  user = User.changeset(%User{}, %{name: "John", email: "john@example.com"})
+  |> Repo.insert!
+
+  response = conn
+  |> get(user_path(conn, :show, user.id))
+  |> json_response(200)
+
+  expected = %{ "data" => %{ "name" => "John", "email" => "john@example.com" } }
+
+  assert response == expected
+end
+```
+
+This is very similar to our `index/2` test, except `show/2` requires a user id, and our data is a single JSON object instead of an array.
+
+When we run our test tells us we need a `show/2` action.
+
+```elixir
+defmodule HelloPhoenix.UserController do
+  use HelloPhoenix.Web, :controller
+
+  alias HelloPhoenix.{User, Repo}
+
+  def index(conn, _params) do
+    users = Repo.all(User)
+    render conn, "index.json", users: users
+  end
+
+  def show(conn, %{"id" => id}) do
+    case Repo.get(User, id) do
+      user -> render conn, "show.json", user: user
+    end
+  end
+end
+```
+
+You'll notice we only handle the case where we successfully find a user. When we TDD we only want to write enough code to make the test pass. We'll add more code when we get to the error handling test for `show/2`.
+
+Running the test tells us we need a `render/2` function that can pattern match on `"show.json"`:
+
+```elixir
+defmodule HelloPhoenix.UserView do
+  use HelloPhoenix.Web, :view
+
+  def render("index.json", %{users: users}) do
+    %{data: render_many(users, HelloPhoenix.UserView, "user.json")}
+  end
+
+  def render("show.json", %{user: user}) do
+    %{data: render_one(user, HelloPhoenix.UserView, "user.json")}
+  end
+
+  def render("user.json", %{user: user}) do
+    %{name: user.name, email: user.email}
+  end
+
+end
+```
+
+When we run the test again, it passes.
+
+The last item we'll cover is the case where we don't find a user in `show/2`.
+
+Try this one on your own and see what you come up with.  One possible solution will be given below.
+
+Walking through our TDD steps, we add a test supplies a non existent user id to `user_path` which returns a 404 status and an error message:
```

Comment:
Think we are missing a word: `we add a test supplies a non existent...` should be `we add a test that supplies a non...`